pkg/core/storage/memcached_store.go

```diff
@@ -83,21 +90,98 @@ func (s *MemCachedStore) GetBatch() *MemBatch {
 
 // Seek implements the Store interface.
 func (s *MemCachedStore) Seek(key []byte, f func(k, v []byte)) {
+	// Create memory store `mem` and `del` snapshots not to hold the lock.
+	memSnapshot := make([]KeyValue, 0)
+	delSnapshot := make(map[string]bool)
+	sk := string(key)
 	s.mut.RLock()
-	defer s.mut.RUnlock()
-	s.MemoryStore.seek(key, f)
-	s.ps.Seek(key, func(k, v []byte) {
-		elem := string(k)
-		// If it's in mem, we already called f() for it in MemoryStore.Seek().
-		_, present := s.mem[elem]
-		if !present {
-			// If it's in del, we shouldn't be calling f() anyway.
-			_, present = s.del[elem]
+	for k, v := range s.MemoryStore.mem {
+		if strings.HasPrefix(k, sk) {
+			memSnapshot = append(memSnapshot, KeyValue{
+				Key:   []byte(k),
+				Value: slice.Copy(v),
+			})
 		}
-		if !present {
-			f(k, v)
+	}
+	for k, v := range s.MemoryStore.del {
+		if strings.HasPrefix(k, sk) {
+			delSnapshot[k] = v
 		}
+	}
+	s.mut.RUnlock()
+	// Sort memSnapshot items for further comparison with ps items.
+	sort.Slice(memSnapshot, func(i, j int) bool {
+		return bytes.Compare(memSnapshot[i].Key, memSnapshot[j].Key) < 0
 	})
+
+	var (
+		data1   = make(chan KeyValue)
+		data2   = make(chan KeyValue)
+		seekres = make(chan KeyValue)
+	)
+	// Seek over memory store.
+	go func() {
+		for _, kv := range memSnapshot {
+			data1 <- kv
+		}
+		close(data1)
+	}()
+
+	// Seek over persistent store.
+	go func() {
+		n := len(memSnapshot)
+		s.mut.Lock()
+		s.ps.Seek(key, func(k, v []byte) {
+			// If it's in mem, we already called f() for it during memSnapshot iteration.
+			i := sort.Search(n, func(i int) bool {
+				return bytes.Compare(memSnapshot[i].Key, k) >= 0
+			})
```

Comment:
`Seek` iterates in order and `memSnapshot` is sorted, so we may maintain a counter instead of a binary search for every key.

Reply:
Are you sure? `memSnapshot` may contain items that are not yet in persistent storage, so I think that we may just restrict the left bound of `memSnapshot` using this counter, but I don't see how can we completely avoid the binary search.

Reply:
Got it, you're right.

Reply:
In fact, we don't even need indexes, we're OK with just a map of keys.